Ensure strict failover/switchover definition difference (#2784)
zalando/patroni: docs/rest_api.rst

Switchover and failover endpoints
---------------------------------

.. _switchover_api:

Switchover
^^^^^^^^^^

> ``/switchover`` endpoint only works when the cluster is healthy (there is a leader). It allows to schedule a switchover at a given time.

Reviewer: ```suggestion
``/switchover`` endpoint only works when the cluster is healthy (there is a leader). It also allows to schedule a switchover at a given time.
```

> When calling ``/switchover`` endpoint candidate can be specified but is not required, in contrast to ``/failover`` endpoint. If candidate is not provided, all the healthy nodes that are allowed to failover participate in the leader race.

Author: maybe you suggest a new version of the sentence then? I want it to be easy to understand:)

> When calling ``/switchover`` endpoint candidate can be specified but is not required, in contrast to ``/failover`` endpoint. If candidate is not provided, all the healthy nodes that are allowed to failover participate in the leader race.

Reviewer 2: I put a suggestion above. I was just wondering if the suggestion made sense or not 😆 

The thing that I found a bit weird was "allowed to failover participate".

> When calling ``/switchover`` endpoint candidate can be specified but is not required, in contrast to ``/failover`` endpoint. If candidate is not provided, all the healthy nodes that are allowed to failover participate in the leader race.

Author: Ok, got it. Well, the idea of the sentence is: 
only the nodes that are:
- healthy (don't lag, are reachable, don't have watchdog failed if one requested)
- allowed to promote (have no limitations like 'nofailover' tag, are in the /sync key if in sync mode)

Maybe then
`If candidate is not provided, all the healthy nodes of the cluster that are allowed to promote participate in the leader race.`

This also makes me think we could have two separate methods for "health" and "allowance" evaluations but it is more for the refactoring

> When calling ``/switchover`` endpoint candidate can be specified but is not required, in contrast to ``/failover`` endpoint. If candidate is not provided, all the healthy nodes that are allowed to failover participate in the leader race.

Reviewer: What about the following?
```suggestion
When calling ``/switchover`` endpoint a candidate can be specified but is not required, in contrast to ``/failover`` endpoint. If a candidate is not provided, all eligible nodes of the cluster will participate in the leader race after the leader stepped down.
```

> In the JSON body of the ``POST`` request, you must specify at least the ``leader`` field and, optionally, the ``candidate`` and ``scheduled_at`` field if you want to schedule a switchover at a specific time.

Reviewer: ```suggestion
In the JSON body of the ``POST`` request, you must specify the ``leader`` field, optionally the ``candidate``, and optionally the ``scheduled_at`` field if you want to schedule a switchover at a specific time.
```

Author: don't fully get it. you want to emphasise that both are optional?

Reviewer: Yes, they are both optional and I want to emphasize it.

**Example:** perform a switchover to a specific node

.. code-block:: bash

	$ curl -s http://localhost:8008/switchover -XPOST -d \
		'{"leader":"postgresql1","candidate":"postgresql2"}'
	Successfully switched over to "postgresql2"

**Example:** schedule a switchover from the leader to any other healthy standby in the cluster at a specific time.

.. code-block:: bash

	$ curl -s http://localhost:8008/switchover -XPOST -d \
		'{"leader":"postgresql0","scheduled_at":"2019-09-24T12:00+00"}'
	Switchover scheduled

Failover
^^^^^^^^

``/failover`` endpoint can be used to perform a manual failover when there are no healthy nodes (e.g. to an asynchronous standby if all synchronous standbys are not healthy enough to promote). However there is no requirement for a cluster not to have leader - failover can also be run on a healthy cluster.

In the JSON body of the ``POST`` request you must specify ``candidate`` field. If ``leader`` field is specified, switchover is triggered. 

**Example:**

.. code-block:: bash

	$ curl -s http://localhost:8008/failover -XPOST -d '{"candidate":"postgresql1"}'
	Successfully failed over to "postgresql1"

.. warning::

> 	:ref:`Be very careful <failover_healthcheck>` using this endpoint, as this can cause data loss in certain situations. In most cases, :ref:`the switchover endpoint <switchover_api>` satisfies the administrator's needs. 

Author: on? isn't it "when"/"about"?

> 	:ref:`Be very careful <failover_healthcheck>` using this endpoint, as this can cause data loss in certain situations. In most cases, :ref:`the switchover endpoint <switchover_api>` satisfies the administrator's needs. 

Reviewer 2: Right, I thought of "when" and typed "on" -- #BlameIsrael

``POST /switchover`` and ``POST /failover`` endpoints are used by ``patronictl switchover`` and ``patronictl failover``, respectively.

``DELETE /switchover`` is used by ``patronictl flush <cluster-name> switchover``.

.. list-table:: Failover/Switchover comparison
   :widths: 25 25 25
   :header-rows: 1

   * -
     - Failover
     - Switchover
   * - Requires leader specified
     - no
     - yes
   * - Requires candidate specified
     - yes
     - no
   * - Can be run in pause
     - yes
     - yes (only to a specific candidate)
   * - Can be scheduled
     - no
     - yes (if not in pause)

.. _failover_healthcheck:

Healthy standby
^^^^^^^^^^^^^^^

> There are a couple of checks that a member of a cluster should pass to be able to participate in the leader race during a switchover or to become a leader as a failover/switchover candidate:

Reviewer: > I know the primary is not well but Patroni doesn't see it

It doesn't fully describe the situation.
Usually Patroni knows that there is no primary, but sometimes it can't do anything about it because there are no members that satisfy all conditions to be automatically selected as a new primary.

- be reachable via Patroni API;
- not have ``nofailover`` tag set to ``true``;
- have watchdog fully functional (if required by the configuration);

> - in case of a switchover or a failover in a healthy cluster, not exceed maximum replication lag (``maximum_lag_on_failover`` :ref:`configuration parameter <dynamic_configuration>`);

Reviewer: ```suggestion
- in case of a switchover in a healthy cluster or automatic failover, not exceed maximum replication lag (``maximum_lag_on_failover`` :ref:`configuration parameter <dynamic_configuration>`);
```

> - in case of a switchover or a failover in a healthy cluster, not have a timeline number smaller than the cluster timeline;

Reviewer: ```suggestion
- in case of a switchover in a healthy cluster or automatic failover, not have a timeline number smaller than the cluster timeline if ``check_timeline`` :ref:`configuration parameter <dynamic_configuration>` is set to ``true``;
```

> - in :ref:`synchronous mode <synchronous_mode>`:

Author: can not find the proper punctuation rule for that, wanted to ask @matthbakeredb, actually:)

> - in :ref:`synchronous mode <synchronous_mode>`:

Reviewer 2: Ah, I see. As Matt lives in England, maybe he knows better how that works in English 😆 
What I mentioned above is very usual in Portuguese -- people write even laws that way.
I could find a reference about that here too -- https://www.grammarbook.com/blog/semicolons/semicolon-use-in-lists/#:~:text=To%20do%20this,of%20each%20line.
But I'm not sure how reliable that reference is 😆

> - in :ref:`synchronous mode <synchronous_mode>`:

Reviewer 3: Haha, being English is not a qualification of the proper use of the language! We're generally terrible at it. However, when I first saw the suggestion it seemed odd to me, at first. I think that it makes sense where each item is a continuation of an initial sentence and breaks something that could have sub-lists within the sentence. Like how it's shown in the example.

This consists a list of related things:

* this one and this one, and another; or
* these, and this one too.

I also think that commas and periods are fine, with the latter more suitable where the line forms a full sentence. If there is one word or phrase in the list it probably would be fine with it being a comma. However, this is informal language, as the link Israel provided mentions it being for formal text, it might be more appropriate to use a semicolon.

In short, I am not 100% sure on the use of the semicolon, I wouldn't get too hung up on it though. I don't have a problem with this suggestion.

> - in :ref:`synchronous mode <synchronous_mode>`:

Reviewer 3: Perhaps we should employ a grammar and style linter so that we have an automated rules system to save us from needing to worry about it. Like vale https://github.com/errata-ai/vale/#usage

> - in :ref:`synchronous mode <synchronous_mode>`:

Reviewer 3: This is a good resource https://docs.gitlab.com/ee/development/documentation/testing.html#vale

> - in :ref:`synchronous mode <synchronous_mode>`:

Reviewer 3: With the gitlab style guide vale reports this for the docs/ directory after running sphinx (so it includes the api docs)

```
67 errors, 2197 warnings, and 511 suggestions found.
```

> - in :ref:`synchronous mode <synchronous_mode>`:

Author: yes, was also thinking of adopting some linter for that, thanks for the link - will take a look

  - In case of a switchover (both with and without a candidate): be listed in the ``/sync`` key members;
  - For a failover in both healthy and unhealthy clusters, this check is omitted.

.. warning::

>     In case of a failover in a cluster without a leader, a candidate will be allowed to promote even if:

Reviewer: ```suggestion
    In case of a manual failover in a cluster without a leader, a candidate will be allowed to promote even if:
```

	- it is not in the ``/sync`` key members when synchronous mode is enabled;
	- its lag exceeds the maximum replication lag allowed;

> 	- it has the timeline number smaller than the cluster timeline.

Reviewer: ```suggestion
	- it has the timeline number smaller than the last known cluster timeline.
```
